archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,299 @@
+  rw le_sqrt at h₁,
+  ring_exp at h₁, norm_num at h₁,
+  norm_cast at h₁,
+  linarith,
+  all_goals { norm_cast, linarith },
+end
+
+
+lemma radical_inequality {n : ℕ} (h : 107 ≤ n) : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3) :=
+begin
+  rw sqrt_le_iff,
+  split,
+  { norm_num,
+    rw le_sqrt,
+    all_goals { norm_cast, linarith } },
+  { ring_exp,
+    rw [pow_two, ← sqrt_mul, sqrt_mul_self],
+    suffices : 24 * sqrt (1 + n) ≤ 2 * n + 36,
+    { linarith },
+    rw mul_self_le_mul_self_iff,
+    ring_exp,
+    rw [pow_two, ← sqrt_mul, sqrt_mul_self],
+    -- Not splitting into cases lead to a deterministic timeout on my machine
+    by_cases p: n ≥ 108,
+    { norm_cast,
+      nlinarith },
+    { simp only [not_le] at p,
+      have : n = 107 := nat.eq_of_le_of_lt_succ h p,
+      subst this,
+      norm_num },
+      swap 3,
+    { norm_num,
+      apply sqrt_nonneg },
+      all_goals { norm_cast, linarith } },
+end
+
+-- We will later make use of the fact that there exists (l : ℕ) such that
+-- n ≤ 2 * l * l - 4 * l and 2 * l * l + 4 * l ≤ 2 * n for n ≥ 107.
+lemma exists_numbers_in_interval (n : ℕ) (hn : 107 ≤ n): ∃ (l : ℕ),
+  (n + 4 * l ≤ 2 * l * l ∧ 2 * l * l + 4 * l ≤ 2 * n) :=
+begin
+  suffices : ∃ (l : ℕ), 2 + sqrt (4 + 2 * n) ≤ 2 * (l : ℝ) ∧ (l : ℝ) ≤ sqrt (1 + n) - 1,
+  { cases this with l t,
+    use l,
+    exact ⟨lower_bound n l t.1, upper_bound n l t.2⟩ },
+  lift floor (sqrt (1 + n) - 1) to ℕ with x,
+  have hx : (x : ℝ) = ⌊sqrt (1 + ↑n) - 1⌋,
+  { norm_cast,
+    convert h,
+    push_cast,
+    norm_num },
+  refine ⟨x, _, _⟩,
+  { suffices : 2 + sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 2),
+    { apply this.trans _,
+      simp only [mul_le_mul_left, zero_lt_bit0, zero_lt_one],
+      rw hx,
+      suffices : sqrt (1 + n) - 1 ≤ ⌊sqrt (1 + n) - 1⌋ + 1,
+      { linarith },
+      have t :  (⌈sqrt (1 + n) - 1⌉:ℝ) ≤ ⌊sqrt (1 + n) - 1⌋ + 1,
+      { norm_cast,
+        exact ceil_le_floor_add_one _ },
+      apply le_trans _ t,
+      exact le_ceil (sqrt (1 + n) - 1) },
+      suffices : sqrt (4 + 2 * n) ≤ 2 * (sqrt (1 + n) - 3),
+      { linarith },
+    exact radical_inequality hn },
+  { rw hx,
+    exact floor_le _ },
+  { rw [floor_nonneg, le_sub, sub_zero, le_sqrt];
+    { norm_cast, linarith } },
+end
+
+lemma exists_triplet_summing_to_squares (n : ℕ) (hn : 100 ≤ n):
+  (∃ (a b c : ℕ), n ≤ a ∧ a < b ∧ b < c ∧ c ≤ 2 * n ∧ (∃ (k : ℕ), a + b = k * k) ∧
+  (∃ (l : ℕ), c + a = l * l) ∧ (∃ (m : ℕ), b + c = m * m)) :=
+begin
+  -- If n ≥ 107, we do not explicitly construct the triplet but use an existence
+  -- argument from lemma above.
+  by_cases p : 107 ≤ n,
+  { have h := exists_numbers_in_interval n p,
+    cases h with l hl,
+    by_cases p : 1 < l,
+    { have h₁ : 4 * l ≤ 2 * l * l, { linarith },
+      have h₂ : 1 ≤ 2 * l, { linarith },
+      refine ⟨2 * l * l - 4 * l, 2 * l * l + 1, 2 * l * l + 4 * l,
+      ⟨_,⟨_,⟨_,⟨_,⟨2 * l - 1, _⟩,⟨2 * l, _⟩,2 * l + 1, _⟩⟩⟩⟩⟩,
```

Comment:
Do you really need to nest all these `⟨⟩`s? I'd guess that you can do something similar to line 151